emukit/examples/simple_gp_model/simple_gp_model.py

```diff
@@ -0,0 +1,121 @@
+        """
+        def optimize_fcn(x):
+            # take exponential to ensure positive values
+            x = np.exp(x)
+            self.lengthscale = x[0]
+            self.kernel_variance = x[1]
+            self.likelihood_variance = x[2]
+            return self._negative_marginal_log_likelihood()
+
+        scipy.optimize.minimize(optimize_fcn, np.log(np.array([self.lengthscale,
+                                                               self.kernel_variance,
+                                                               self.likelihood_variance])))
+
+    def predict(self, x_new: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Predict from model
+
+        :param x_new: (n_points, n_dims) array containing points at which the predictive distributions will be computed
+        :return: Tuple containing two (n_points, 1) arrays representing the mean and variance of the predictive
+                 distribution at the specified input locations
+        """
+        K = self._calc_kernel(self.x)
+        idxs = list(range(self.x.shape[0]))
+        K[idxs, idxs] += self.jitter + self.likelihood_variance
+        L = np.linalg.cholesky(K)
+
+        K_xs = self._calc_kernel(self.x, x_new)
+
+        tmp = scipy.linalg.solve_triangular(L, K_xs, lower=True)
+        tmp2 = scipy.linalg.solve_triangular(L, self.y, lower=True)
+
+        mu = np.dot(tmp.T, tmp2)
+        var = (self.kernel_variance - np.sum(np.square(tmp), axis=0))[:, None]
+        return mu, var
+
+    def _calc_kernel(self, X, X2=None):
+        """
+        Implements an RBF kernel with no ARD
+        """
+        if X2 is None:
+            Xsq = np.sum(np.square(X), 1)
+
+            XXt = np.dot(X, X.T)
+            r2 = -2. * XXt + (Xsq[:, None] + Xsq[None, :])
+            r2 = np.clip(r2, 0, np.inf)
+        else:
+            X1sq = np.sum(np.square(X), 1)
+            X2sq = np.sum(np.square(X2), 1)
+            r2 = -2. * np.dot(X, X2.T) + (X1sq[:, None] + X2sq[None, :])
+            r2 = np.clip(r2, 0, np.inf)
+
+        return self.kernel_variance * np.exp(-r2 / self.lengthscale ** 2)
+
+    def _negative_marginal_log_likelihood(self) -> float:
+        """
+        Negative marginal log likelihood
+        """
+        K = self._calc_kernel(self.x)
+
+        # Add some jitter to the diagonal
+        idxs = list(range(self.x.shape[0]))
+        K[idxs, idxs] += self.jitter + self.likelihood_variance
+
+        # cholesky decomposition of covariance matrix
+        L = np.linalg.cholesky(K)
+
+        # Log determinant of the covariance matrix
+        log_det = 2. * np.sum(np.log(np.diag(L)))
+
+        # calculate y^T K^{-1} y using the cholesky of K
+        tmp = scipy.linalg.solve_triangular(L, self.y, lower=True)
+        alpha = scipy.linalg.solve_triangular(L.T, tmp, lower=False)
+        log_2_pi = np.log(2 * np.pi)
+        return -0.5 * (-self.y.size * log_2_pi - self.y.shape[1] * log_det - np.sum(alpha * self.y))
```

Comment:
do you mean self.x.size here instead of self.y.size?

Reply:
I think it should be y.size as the nmll only depends on the dimensionality of the output rather than input

Reply:
yes, totally. Not sure what I was thinking o_0